coroutine: exception: retain exception_ptr type

bhalevy opened the pull request:
Currently, make_exception/return_exception call
std::make_exception_ptr unconditionally.
When called over a std::exception_ptr, std::make_exception_ptr
loses the arg's exception type and creates an untyped
exception_ptr.

Instead, use a templated constructor for struct exception
that just moves the given exception_ptr into its member
to retain its type.  It's also a bit more efficient
since it save the construction of a new exception ptr
and one move of it (into the constructor's parameter).

Added a respective unit test.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

piodul commented on include/seastar/coroutine/exception.hh:
```diff
@@ -68,7 +68,11 @@ namespace coroutine {
+    explicit exception(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
+
+    template <typename E>
+    explicit exception(E&& e) noexcept : eptr(std::make_exception_ptr(std::forward<E>(e))) {}
```
Obligatory bikeshed: should this be unconditionally noexcept? `std::make_exception_ptr` takes its argument by value, so we need to move/copy `e` which can throw. I guess crashing might be the most sensible option here, just wanted to point this out in case this was not intentional.

bhalevy replied:
This is in line with other places we don't let low-level exception handling throw.

bhalevy replied:
Ref: https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/include/seastar/core/future.hh#L2086-L2089

piodul replied:
I see, thanks.

bhalevy commented:
@piodul can you please add `log_exception_trace` trace calls on these paths too, like we do for `future::make_exception_future`?

bhalevy commented:
I opened https://github.com/scylladb/seastar/issues/1091 for adding `log_exception_trace`.

piodul commented:
> @piodul can you please add `log_exception_trace` trace call to the `exception(T&& e) path`, like we do for `future::make_exception_future(Exception&& ex)`?

OK, I'll try to send a PR soon.

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -68,7 +68,11 @@ namespace coroutine {
+    explicit exception(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
+
+    template <typename E>
+    explicit exception(E&& e) noexcept : eptr(std::make_exception_ptr(std::forward<E>(e))) {}
```
It's a bad idea to allow construction from both exception_ptr and any type. A small mistake (like exception(&ep)) can complete change the behavior (seastar::make_exception_future has the same problem).

Let's use a different container for it.

avikivity replied:
Or a helper function. In fact why not let the caller call make_exception_ptr?

avikivity replied:
And looking at the patch changelog, perhaps delete make_exception signatures that take exception_ptr.

bhalevy replied:
One reason not to accept exception objects and not require the caller to call make_exception_ptr is to trace them,
as in https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/include/seastar/core/future.hh#L2087
Since they are never thrown (where they are traced by https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/src/core/exception_hacks.cc#L146).

We use a template instead of accept a reference to the std::exception in order to not lose the exception type when calling `std::make_exception_ptr`.  But I can add `requires std::derived_from<std::remove_reference_t<E>, std::exception>` to restrict the usage.

bhalevy replied:
@avikivity if we change the api to accept only exception_ptr or classes derived from std::exception, we'll need an api version change since this will break use cases like https://github.com/scylladb/seastar/blob/1424d34c93846c8e23d60e7550f39d621643ac8e/tests/unit/smp_test.cc#L37-L59

bhalevy replied:
Keeping the restrictions only for coroutine::{return,make}_exception should be less of a problem since I image they have much fewer users.

bhalevy commented:
In v2 (edc0e2924f440d639c0b56184c80cbc3483ab63c):
- added constraints to limit the usage of coroutine::{make,return}_exception to exception / exception_ptr types
- added call to `log_exception_trace`, fixing #1091

bhalevy commented:
ping, please review.

psarna commented:
Looks good, but existing reviewers should ack first. Also, am I correct that the new constraints are technically not backward compatible, but we cross our fingers that this API is still rather fresh?

psarna commented:
oh no, wait, they still look compatible - where we only used exception_ptr before, we now also accept certain classes of exceptions directly

bhalevy commented:
> oh no, wait, they still look compatible - where we only used exception_ptr before, we now also accept certain classes of exceptions directly

Previously, the template accepted anything, exception / exception_ptr, or just any value and turned it into an exception_ptr using `std::make_exception_ptr(std::forward<T>(t))`.

The gist of this issue is that if `t` is already an exception_ptr, make_exception_ptr of an exception_ptr buries the original exception_ptr down in its belly, hiding its type.

piodul commented on include/seastar/coroutine/exception.hh:
```diff
@@ -87,9 +93,10 @@ struct exception {
-exception make_exception(T&& t) {
-    return exception(std::make_exception_ptr(std::forward<T>(t)));
+exception make_exception(T&& t) noexcept {
+    return exception(std::forward<T>(t));
```
The `make_exception`/`return_exception` functions don't seem to accept `std::exception_ptr` by copy. For example, this doesn't compile for me:

```c++
    std::exception_ptr eptr;
    (void)make_exception(eptr);
```

Is this intentional? I guess there might be legitimate cases where copying is preferred.

bhalevy replied:
I wanted to save an extraneous move, if the constructor was defined as exception(std::exception_ptr). The compiler doesn't seem to optimize it well as far as @denesb observed.  But there was no intention of disallowing passing a value by copy.
I'll fix that in the next version.

psarna replied:
I think we do want to disallow it. The rationale is that copying an exception_ptr is costly, because it touches its atomic reference counts. On the other hand, if somebody really wants to pass a copy (I don't really know why though), they can do so explicitly by creating a copy and moving it.

bhalevy replied:
@psarna it is quite cumbersome to copy an exception_ptr explicitly since make_exception_ptr(ex), misleading as it is, will mask the original exception ptr, so one needs to assign the exception_ptr to copy into a new variable and then move it, yikes... :-\
We do allow copying exception ptrs in many other paths.
As typical use case is an aborted object that keeps the abort reason as an exception_ptr member
and when accessed it returns an exceptional future by copying the exception_ptr.  You don't want to move it in this case since the aborted state is sticky.

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -68,7 +68,13 @@ namespace coroutine {
+
+    template <typename E>
+    requires std::derived_from<std::remove_reference_t<E>, std::exception>
+    explicit exception(E&& e) noexcept : eptr(std::make_exception_ptr(std::forward<E>(e))) {
```
How about: `explicit exception(std::derived_from<std::exception> auto&& e)` to reduce boilerplate.

bhalevy replied:
I'll try that. If it works, I'll push v4.

bhalevy replied:
sone in v4.  thanks for your suggestion

avikivity commented on include/seastar/coroutine/exception.hh:
````diff
@@ -87,9 +93,10 @@ struct exception {
 /// co_return coroutine::make_exception(std::runtime_error("something failed miserably"));
 /// ```
 template<typename T>
+requires std::same_as<T, std::exception_ptr> || std::derived_from<std::remove_reference_t<T>, std::exception>
````
This is just an indication the API is wrong. It behaves differently for different argument types. It should be two separate APIs.

bhalevy replied:
Agreed. Done that way in v3.

bhalevy commented:
In v3 (97e387f5):
- separated entry points between std::exception_ptr input paths and std::exception-derived arg.
- added support for `const std::exception_ptr&` arg and extended the added unit test to cover it.

bhalevy commented:
In v4 (f7a1ee21):
- used `std::derived_from<std::exception> auto&&` to reduce boilerplate as per https://github.com/scylladb/seastar/pull/1090#discussion_r895468047

bhalevy commented on include/seastar/coroutine/exception.hh:
```diff
@@ -68,7 +68,12 @@ namespace coroutine {
+    explicit exception(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
+    explicit exception(const std::exception_ptr& eptr) noexcept : eptr(eptr) {}
+
+    explicit exception(std::derived_from<std::exception> auto&& e) noexcept : eptr(std::make_exception_ptr(std::move(e))) {
```
@nyh it can in theory (well, it could before this change). The question is if and why we want to support it.

nyh replied:
Because C++ is not Java, so exceptions do not **have** to derive from std::exception. It is true that std::exception has benefits (such as `what()` method that our operator<<(std::exception_ptr) can use), but since C++ doesn't absolutely require std::exception, I think neither should we. Especially since in the past we didn't, so why start now?

bhalevy replied:
@avikivity what's your take on this?
I can bring back support for make_exception_ptr on anything that's not an exception_ptr including, but not limiting to std::exception's.  This would still fix the original issue of masking the original exception_ptr and maintain compatibility with `throw` and `make_exception_future`.  I don't like carrying non-exception-type values in exceptional futures, but the language supports it... :-\

nyh replied:
@avikivity that's a different issue. Even if we didn't treat exception_ptr differently from other types, my question is why require that those other types must inherit from std::exception - since this is not a C++ requirement.

bhalevy commented on include/seastar/coroutine/exception.hh:
```diff
@@ -104,10 +117,19 @@ exception make_exception(T&& t) {
+
+[[nodiscard]]
+exception return_exception(std::derived_from<std::exception> auto&& e) noexcept {
+    return exception(std::move(e));
```
> Isn't std::forward actually needed because this is a template, not an rvalue reference?

Yes, taking the template <typename E> out, we cannot std::forward<E>(e), just std::move it.
I don't think this makes any difference in performance since e is passed be rvalue reference, not by value, so its move constructor or assignment operator are not called until it lands at its destination.

nyh replied:
Isn't "auto&&" a template?
In other words, isn't this && a sign of a **universal reference**, _not_ an **rvalue reference**?
A quick Google search yielded https://stackoverflow.com/questions/13230480/what-does-auto-tell-us which suggests maybe it's not an rvalue.

bhalevy replied:
Github took the liberty to hide stuff in angular brackets in my comment above, so pasting it below with hopefully correct formatting:

> Yes, taking the template `<typename E>` out, we cannot `std::forward<E>(e)`, just std::move it.

so, I don't think we can use [std::forward](https://en.cppreference.com/w/cpp/utility/forward) here.

nyh replied:
But what I tried to reply that "auto&&" is just another syntax for a template. It's a new (almost) C++20 syntax called "abbreviated function template" - see for example https://ggulgulia.medium.com/c-20-concepts-part-4-389a7649f084

Looking above, it seems Avi asked you to change the template to auto&& just to have shorter syntax - I don't think it changes anything.

bhalevy replied:
ok, got you.  We can indeed use `std::forward<decltype(e)>` here.

nyh replied:
Wow, this looks horrible - i.e., it looks like real C++ ;-)

bhalevy commented on tests/unit/coroutines_test.cc:
```diff
@@ -466,6 +466,20 @@ SEASTAR_TEST_CASE(test_coroutine_exception) {
 
+SEASTAR_TEST_CASE(test_coroutine_return_exception_ptr) {
+    co_await check_coroutine_throws<std::runtime_error>([] (int& counter) -> future<> {
+        co_await coroutine::return_exception(std::runtime_error("threw"));
```
I agree we can make it work, but I don't know why want to support that.

avikivity commented:
> In v3 ([97e387f](https://github.com/scylladb/seastar/commit/97e387f5e9e2690175d07ac83a36701740e9d3b7)):
> 
> * separated entry points between std::exception_ptr input paths and std::exception-derived arg.

`exception` struct still has a constructor from non-exception_ptr.

> * added support for `const std::exception_ptr&` arg and extended the added unit test to cover it.

bhalevy commented:
> > In v3 ([97e387f](https://github.com/scylladb/seastar/commit/97e387f5e9e2690175d07ac83a36701740e9d3b7)):
> > 
> > * separated entry points between std::exception_ptr input paths and std::exception-derived arg.
> 
> `exception` struct still has a constructor from non-exception_ptr.

@avikivity based on the discussion with @nyh, I think we should separate the bug fix - that is masking out the exception_ptr type, from hardening the interface to prevent unintentional usage.

Would you just like `coroutines::exception` to accept only `std::exception_ptr` and do the conversion in `return_exception`?

bhalevy commented:
In v6 (ee5c62fe):
- kept the calls to `make_exception_ptr` on at the `{make,return}_exception` helpers.
  - `coroutine::exception` constructors get only `std::exeption_ptr&&` or `const std::exeption_ptr&`
- separated support of copying the exception_ptr via a const& into a standalone patch

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -30,7 +30,7 @@ namespace internal {
     std::exception_ptr eptr;
 
-    explicit exception_awaiter(std::exception_ptr&& eptr) : eptr(std::move(eptr)) {}
+    explicit exception_awaiter(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
```
I don't understand what's an untyped exception_ptr.

bhalevy replied:
The symptom is that the boost testing framework reports the following:
```
unknown location(0): fatal error: in "sstable_directory_test_table_scan_invalid_file": unknown type
```

I guess that the type of the held object is `std::exception_ptr` (rather than untyped as I said)

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -68,7 +68,7 @@ namespace coroutine {
 struct exception {
     std::exception_ptr eptr;
-    explicit exception(std::exception_ptr eptr) : eptr(std::move(eptr)) {}
+    explicit exception(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
```
What happens if someone supplies a const exception_ptr? Why does that deserve a compilation failure?

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -104,9 +111,16 @@ exception make_exception(T&& t) {
+template<typename T>
+requires std::same_as<std::remove_reference_t<T>, std::exception_ptr>
+[[nodiscard]]
+exception return_exception(T&& t) noexcept {
```
It's not good form to have an API accept both std::exception_ptr and a value type and treat them differently.

bhalevy replied:
Do you suggest to get only `std::exception_ptr` and let the caller make it if needed?
We can do that, but I'm afraid that will require to bump the api version.

bhalevy replied:
Also, do we want to present narrower API than make_exception_future(), or for the sake of symmetry we should keep up with the badly formed API, which I don't find so bad by the way.
To me it feels like many other polymorphic interfaces in the c++ standard library that accept different input types and treat them differently.  I don't really understand why here you find it so distasteful.

avikivity commented on include/seastar/coroutine/exception.hh:
```diff
@@ -97,6 +98,7 @@ exception make_exception(T&& t) noexcept {
 [[nodiscard]]
 exception make_exception(T&& t) noexcept {
+    log_exception_trace();
     return exception(std::make_exception_ptr(std::forward<T>(t)));
```
I don't understand how we can know that we want logging so deep in system code.

bhalevy replied:
Exactly like in https://github.com/scylladb/seastar/blob/23d28aba656b869367122fb83f0ccd3486fc5ae4/include/seastar/core/future.hh#L2086-L2088

We want to trace where the exception was generated and the assumption is that this is one place where the user created an exception object that is passed on up the stack, rather than being thrown.

For the throw case, we have https://github.com/scylladb/seastar/blob/23d28aba656b869367122fb83f0ccd3486fc5ae4/src/core/exception_hacks.cc#L138-L149

avikivity replied:
Ah, trace level makes if less terrible.

nyh commented:
@bhalevy @avikivity after this long thread, I'm completely confused what is actually intended here. I just wrote the following code in Scylla, to "rethrow" an exception from a coroutine:

```c++
       co_return coroutine::make_exception(std::move(eptr));
```
to replace the traditional C++ code:

```c++
     std::rethrow_exception(eptr);
```

Is this a good idiom, or bad idiom? Is this patch fixing this idiom, changing it, or what?

By the way, this example makes me think of two things:

1. The version of make_exception which takes an exception_ptr (which @avikivity asked for in the above discussion, if I understood correctly) should be called **rethrow_exception**.
2. This new rethrow_exception should take the exception_ptr by reference, not rvalue reference, to avoid the need for the ugly explicit std::move() in the call.

bhalevy commented:
> @bhalevy @avikivity after this long thread, I'm completely confused what is actually intended here. I just wrote the following code in Scylla, to "rethrow" an exception from a coroutine:
> 
> ```c++
>        co_return coroutine::make_exception(std::move(eptr));
> ```
> 
> to replace the traditional C++ code:
> 
> ```c++
>      std::rethrow_exception(eptr);
> ```
> 
> Is this a good idiom, or bad idiom?

good, IMO

> Is this patch fixing this idiom, changing it, or what?

It is fixing it.
The current code wraps `eptr` in an extraneous std::exception_ptr.
This patch eliminates it.

> 
> By the way, this example makes me think of two things:
> 
> 1. The version of make_exception which takes an exception_ptr (which @avikivity asked for in the above discussion, if I understood correctly) can be called **coroutine::rethrow_exception()**.

I disagree, since it's not rethrown.
On the contrary, the whole point of the coroutine::make_exception (or return_exception that are both now doing the same) is to avoid rethrowing the exception as a performance optimization.

> 2. This new rethrow_exception should take the exception_ptr by reference, not rvalue reference, to avoid the need for the ugly explicit std::move() in the call.

But then it will always have to copy it, which is more expensive, doesn't it?

nyh commented:
> > @bhalevy @avikivity after this long thread, I'm completely confused what is actually intended here. I just wrote the following code in Scylla, to "rethrow" an exception from a coroutine:
> > ```c++
> >        co_return coroutine::make_exception(std::move(eptr));
> > ```    
> > to replace the traditional C++ code:
> > ```c++
> >      std::rethrow_exception(eptr);
> > ```
> >     
> > Is this a good idiom, or bad idiom?
> 
> good, IMO

Does @avikivity share your opinion here? Because I got a feeling from his review (but maybe I misunderstood?) that he wanted a different function name, not make_exception() for this use case.

> > 1. The version of make_exception which takes an exception_ptr (which @avikivity asked for in the above discussion, if I understood correctly) can be called **coroutine::rethrow_exception()**.
> 
> I disagree, since it's not rethrown. On the contrary, the whole point of the coroutine::make_exception (or return_exception that are both now doing the same) is to avoid rethrowing the exception as a performance optimization.

Yes, maybe the name "rethrow" is a bit misleading / ambiguous. On the flip side if I understood Avi correctly, he wants another name for this function, and I don't know what name to suggest. "rethrow_exception" looked like a name that people will remember, but perhaps you're right and this specific name is not a good idea.

> 
> > 2. This new rethrow_exception should take the exception_ptr by reference, not rvalue reference, to avoid the need for the ugly explicit std::move() in the call.
> 
> But then it will always have to copy it, which is more expensive, doesn't it?

I didn't suggest to take exception_ptr by value (requiring a copy), but by reference - "exception_ptr&" (not "const"). I think this can also avoid a copy, without forcing the user to type out "std::move".

bhalevy commented:
> I didn't suggest to take exception_ptr by value (requiring a copy), but by reference - "exception_ptr&" (not "const"). I think this can also avoid a copy, without forcing the user to type out "std::move".

@nyh eventually we keep a std::exception_ptr member in class coroutine::exception.
The current version of this series offers 2 constructors with: exception_ptr&& and const exception_ptr&.
In the first variant, the input is moved, and in the second, it is copied into the exception::eptr member.
You suggestion will cause a copy (unless we can get rid of class coroutine::exception entirely).

I think that Avi's direction (and please clarify, @avikivity) is to support only exception(std::exception_ptr)
and then it is up to the user to move, copy, or make_exception_ptr for whatever payload they want to pass back.

nyh commented:
> > I didn't suggest to take exception_ptr by value (requiring a copy), but by reference - "exception_ptr&" (not "const"). I think this can also avoid a copy, without forcing the user to type out "std::move".
> 
> @nyh eventually we keep a std::exception_ptr member in class coroutine::exception. The current version of this series offers 2 constructors with: exception_ptr&& and const exception_ptr&. In the first variant, the input is moved, and in the second, it is copied into the exception::eptr member. You suggestion will cause a copy (unless we can get rid of class coroutine::exception entirely).
> 
> I think that Avi's direction (and please clarify, @avikivity) is to support only exception(std::exception_ptr) and then it is up to the user to move, copy, or make_exception_ptr for whatever payload they want to pass back.

Perhaps we could have `coroutine::exception(eptr)` taking **only** exception_ptr, and then `coroutine::make_exception()` as a shortcut to `coroutine::exception(std::make_exception_ptr(...))?

bhalevy commented:
@avikivity ping

avikivity commented:
> > > @bhalevy @avikivity after this long thread, I'm completely confused what is actually intended here. I just wrote the following code in Scylla, to "rethrow" an exception from a coroutine:
> > > ```c++
> > >        co_return coroutine::make_exception(std::move(eptr));
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > to replace the traditional C++ code:
> > > ```c++
> > >      std::rethrow_exception(eptr);
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Is this a good idiom, or bad idiom?
> > 
> > 
> > good, IMO
> 
> Does @avikivity share your opinion here? Because I got a feeling from his review (but maybe I misunderstood?) that he wanted a different function name, not make_exception() for this use case.

I want different names for exception_ptr and generic object parameters. It's too easy to make a small mistake and use the wrong one.

avikivity commented:
> > I didn't suggest to take exception_ptr by value (requiring a copy), but by reference - "exception_ptr&" (not "const"). I think this can also avoid a copy, without forcing the user to type out "std::move".
> 
> @nyh eventually we keep a std::exception_ptr member in class coroutine::exception. The current version of this series offers 2 constructors with: exception_ptr&& and const exception_ptr&. In the first variant, the input is moved, and in the second, it is copied into the exception::eptr member. You suggestion will cause a copy (unless we can get rid of class coroutine::exception entirely).
> 
> I think that Avi's direction (and please clarify, @avikivity) is to support only exception(std::exception_ptr) and then it is up to the user to move, copy, or make_exception_ptr for whatever payload they want to pass back.

We cannot remove make_exception(), it's already there.

bhalevy commented:
> We cannot remove make_exception(), it's already there.

Of course. What I meant is have only a single explicit constructor for exception that accepts an exception_ptr and make all the conversions via the make_exception (and return_exception that we're also already committed to) interfaces.

The reason I implemented the conversions in the coroutine::exception constructors was not to replicate code between `make_exception` and `return_exception` that are now exactly the same.

> I want different names for exception_ptr and generic object parameters. It's too easy to make a small mistake and use the wrong one.

I see.  How about the following distinction?
`coroutine::make_exception(exception_ptr)`
vs.
`coroutine::make_exception_ptr(T&&)`

bhalevy commented:
@avikivity please see proposal above.
I'd like to get this PR to closure and fix the issue.

avikivity commented:
Wait. The current entry point is `co_await coroutine::return_exception()`. This is good because `return` indicates we're returning even though this is co_await.

So we should add another entry point, `co_await coroutine::return_exception_ptr()`. This also has `return` so it indicates we're returning from the coroutine, and `exception_ptr` clearly describes the input. And there is no clash with existing APIs.

We can make return_exception() reject exception_ptr inputs.

avikivity commented:
`co_await coroutine::return_exception(std::exception_ptr())` will do something wrong today, so no need to deprecate it.

bhalevy commented:
In that case, I'd be in favor of deprecating the `coroutine::make_exception` entry point and leave only return_exception.
Originally we used it to `co_return coroutine::make_exception()` when to coroutine returned a non-void future, but
`co_return coroutine::exception()` works just as well.
I think it would be sufficient to support the `co_await coroutine::return_exception{,_ptr}()` entry points for `future<>` returning coroutines.

bhalevy commented:
I'll push a new version with the above proposal

bhalevy commented:
In v7 (40e140c2)
- cleaned up patchset
- added coroutine::return_exception_ptr
- deprecated coroutine::make_exception

nyh commented on include/seastar/coroutine/exception.hh:
```diff
@@ -30,7 +30,7 @@ namespace internal {
     std::exception_ptr eptr;
 
-    explicit exception_awaiter(std::exception_ptr&& eptr) : eptr(std::move(eptr)) {}
+    explicit exception_awaiter(std::exception_ptr&& eptr) noexcept : eptr(std::move(eptr)) {}
```
Nitpick: although I am guessing exception_ptr is indeed, like you said, nothrow movable in practice, I am not sure this is actually mandated by the standard. Why was it important to add this "noexcept" here? Is it related to the purpose of this series?

bhalevy replied:
Not specifically for the purpose of this series, but in general, with `co_await coroutine::return_exception(some_exception())` you don't want an exception to escape from this "control" statement (although it is not a bug).  This behavior is akin to future exception and continuation handling where we terminate if we fail to create a continuation or make an exceptional future.

nyh commented on include/seastar/coroutine/exception.hh:
````diff
@@ -104,6 +111,11 @@ exception make_exception(T&& t) noexcept {
 /// co_await coroutine::return_exception(std::runtime_error("something failed miserably"));
 /// ```
+[[nodiscard]]
+inline exception return_exception_ptr(std::exception_ptr ex) noexcept {
````
Doesn't this pass-by-value doesn't force the compiler to do an extra copy of the exception_ptr... Since exception_ptr has a non-trivial copy constructor (which needs to deal with the reference count), I don't see how the compiler can avoid generating it...
I would have done std::exception_ptr& and std::exception_ptr&& variants - not pass by value.

nyh replied:
ׁHmm, I think you're right. Ok. The extra move can (and should) be optimized out, so it's fine.

bhalevy replied:
That said, moving an exception_ptr should be cheap (same as a std::shared_ptr I think) as it doesn't copy the exception object.
